### PR TITLE
fix: metadata service respects initial installs from new vscode instances for improved metadata accuracy

### PR DESCRIPTION
**Pull request by hikchoi**
# fix: metadata service respects initial installs from new vscode instances for improved metadata accuracy

This PR:
- fixes an issue where an install event on a fresh vscode instance is recorded as an "initial install" globally.
  - this may have caused problems with prompt logic that rely on the integrity of the metadata service.

# Dendron Extended PR Checklist

### Basics

- [x] code should follow [Code Conventions](https://wiki.dendron.so/notes/773e0b5a-510f-4c21-acf4-2d1ab3ed741e.html)
- [x] circular dependency check: make sure your code is not introducing new circular dependencies in plugin-core.  See [Avoiding Circular Dependencies](https://wiki.dendron.so/notes/pMS27sHxbWeKMoPRrWEzs.html).
- [x] sticking to existing conventions instead of creating new ones 
  - eg: [if configuration for utilities are already in one module or package, add future utilities there as well](https://github.com/dendronhq/dendron/pull/1960#discussion_r786228021)

### Extended

- General
  - [x] check whether code be simplified
  - [x] check if similar function already exist in the codebase. if so, can it be re-used?
  - [x] check if this change adversely impact performance
- Operations
  - [x] when shipping this change, will it just work or will it introduce additional operational overhead due to complicated interface or known bugs?
- Architecture
  - [x] check if code is introducing changes on a foundational class or interface. if so, call for design review if needed 
    - eg: [making changes to DNode](https://github.com/dendronhq/dendron/pull/2158#pullrequestreview-854689586)

### Basics

- [x] if you are adding analytics related changes, make sure you [Document](https://wiki.dendron.so/notes/8ThPaB9iXXm2Szk3C9kFt.html) changes in airtable

### Extended

- [~] can we track the performance of this change to know if it is _successful_? 
  - eg: [see usage for export pod](https://github.com/dendronhq/dendron/pull/2190#pullrequestreview-855715612)

## 

### Basics

- [x] [Write Tests](https://wiki.dendron.so/notes/99q7A73uGmCwu2KvSHZro.html) 
- [x] [Confirm existing tests pass](https://wiki.dendron.so/notes/99q7A73uGmCwu2KvSHZro.html)
- [x] [Confirm manual testing](https://wiki.dendron.so/notes/99q7A73uGmCwu2KvSHZro.html) 
- [x] Common cases tested
- [x] 1-2 Edge cases tested
- [~] If your tests changes an existing snapshot, snapshots have been [updated](https://wiki.dendron.so/notes/99q7A73uGmCwu2KvSHZro.html)

### Extended

- [~] If you are adding a new language feature (graphically visible in VS Code/preview/publishing), an example is included in the [test workspace](https://wiki.dendron.so/notes/dtMsF12SF2SUhLN10sYe2.html)

- CSS
  - [~] display is correct for following dimensions
    - [~] sm: screen ≥ 576px, eg. iphonex, (375x812)
    - [~] lg: screen ≥ 992px
    - [~] xxl: screen ≥ 1600px eg. mac (1600x900)
  - [~] display is correct for following browsers (across the various dimensions)
    - [~] safari
    - [~] firefox
    - [~] chrome

### Basics

- [~] if your change reflects documentation changes, also submit a PR to [dendron-site](https://github.com/dendronhq/dendron-site) and mention the doc PR link in your current PR

## 

### Basics

- [~] does this change introduce a new or better way of doing things that others need to be aware of? if so, an async should be created and a process added in [Development](https://wiki.dendron.so/notes/3489b652-cd0e-4ac8-a734-08094dc043eb.html) or [Packages](https://wiki.dendron.so/notes/32cdd4aa-d9f6-4582-8d0c-07f64a00299b.html)

## 

### Extended

- [~]  is this a developer BREAKING change? if another person cloning from this branch will need to adjust their dependencies or mental model of the architecture, then it is. if this is the case, make sure this is communicated according to [Close Loop](https://wiki.dendron.so/notes/iZ8vpfY0n1E3Nq3IC1Y7X.html)
  - eg. [breaking dev change due to new dependency](https://github.com/dendronhq/dendron/pull/2188#pullrequestreview-855696330)

### CODE REVIEW

File: packages/common-all/src/analytics.ts [hunk 1]
```diff
@@ -3,6 +3,7 @@ export enum VSCodeEvents {
   ServerCrashed = "ServerCrashed",
   InitializeWorkspace = "InitializeWorkspace",
   Install = "Install",
+  NewInstanceInstall = "NewInstanceInstall",
```
Comment (kevinslin) [inline packages/common-all/src/analytics.ts:6]: instead of creating a new event, can we use "Install" event but with the property, "newInstance" set to true?

**Review by kevinslin — CHANGES_REQUESTED** [review]

File: packages/common-all/src/analytics.ts (hunk 1 above)
Comment (hikchoi) [reply packages/common-all/src/analytics.ts]: fixed, and also recorded in airtable

File: packages/plugin-core/src/_extension.ts [hunk 2]
```diff
@@ -394,9 +394,22 @@ export async function _activate(
...
+    // check if this is an install event, but a repeated one on a new instance.
+    let isNewInstanceInstall = false;
+
     // set initial install ^194e5bw7so9g
     if (extensionInstallStatus === InstallStatus.INITIAL_INSTALL) {
-      MetadataService.instance().setInitialInstall();
+      // even if it's an initial install for this instance of vscode, it may not be for this machine.
+      // in that case, we should skip setting the initial install time since it's already set.
+      const metadata = MetadataService.instance().getMeta();
+      if (metadata.firstInstall === undefined) {
```
Comment (jonathanyeung) [inline packages/plugin-core/src/_extension.ts:406]: Any chance that during our transition, that someone who is an active user of Dendron doesn't have `metadata.firstInstall` defined yet? 

My preference would also to be check for the presence of the `dendron.uuid` file - if it exists, then also don't set to initial install.  That way, since the uniqueness of `track` is based on .uuid, we should ensure we're 100% aligned and we won't have old users showing up in the onboarding funnel.

File: packages/plugin-core/src/_extension.ts [hunk 3]
```diff
@@ -805,12 +821,20 @@ async function showWelcomeOrWhatsNew({
...
+        msg: `extension, ${
+          isNewInstanceInstall
+            ? "initial install"
+            : "install on new vscode instance"
+        }`,
+      });
       // track how long install process took ^e8itkyfj2rn3
       AnalyticsUtils.track(VSCodeEvents.Install, {
         duration: getDurationMilliseconds(start),
+        isNewInstanceInstall,
```
Comment (jonathanyeung) [inline packages/plugin-core/src/_extension.ts:835]: Can we rename this to something like "isSecondaryInstallation" or something like that?  Because even if this is the first ever installation, it's still technically a "new instance install".

File: packages/plugin-core/src/test/suite-integ/Extension.test.ts [hunk 4]
```diff
@@ -1093,6 +1093,38 @@ suite("GIVEN Dendron plugin activation", function () {
       }
     );
   });
+
+  describe("AND WHEN fresh install on new vscode instance", () => {
```
Comment (jonathanyeung) [inline packages/plugin-core/src/test/suite-integ/Extension.test.ts:1097]: for clarity, can we reword this to "on a secondary vscode instance"

**Review by jonathanyeung — CHANGES_REQUESTED** [review]

**Comment by kevinslin** [comment]
can you rebase from master? fixes some build issues

File: packages/plugin-core/src/_extension.ts (hunk 2 above)
Comment (hikchoi) [reply packages/plugin-core/src/_extension.ts:406]: > Any chance that during our transition, that someone who is an active user of Dendron doesn't have metadata.firstInstall defined yet?

That's possible. That's why we have a temporary backfill logic that follows right after.
Though if an active user doesn't have it defined yet, and never went through that backfill, and also has installed a fresh instance of vscode, this would fill it up with the current time.

Your suggestion to check for the uuid would solve the problem though. So I'll go with that.

**Comment by hikchoi** [comment]
Added suggested changes. Took some troubleshooting to figure out when the first Segment client instantiation actually happens. Maybe it's time for a deep cleansing of the activation logic 😄

@kevinslin I noticed that you removed an entire suite of tests here: https://github.com/dendronhq/dendron/pull/2714/files#diff-60b981e53d27fcadc5d898d4d161eccda9235211cc64e308cce08114e08b8fa1L987

I'm adding them back as they are needed to test the behavior of my changes, but were there any particular reason those were removed in your PR?

I did have to go back to your changes to remind myself how to clean up the extension context to make it work properly.

**Review by jonathanyeung — APPROVED** [review]
